### ISSUE TITLE
Feature Request: Two-factor authentication (2FA)

### ISSUE THREAD

**Issue by 475392689485** [grocy/grocy#281]
It would be nice to have 2FA for additional security.

**Comment by DeeeeLAN** [comment]
Possible implementation option:
http://hazardedit.com/2017/11/02/implementing-totp-google-authenticator-php/
https://github.com/PHPGangsta/GoogleAuthenticator

**Comment by Braintelligence** [comment]
Is there anything planned on the roadmap about this?

**Comment by jsternadel** [comment]
If you are considering 2fa please take a look at [authelia](https://github.com/authelia/authelia) or at least consider it as an optional 2fa solution to standard TOTP.  The more home networks grow the more important it will be for centralized authentication in my opinion.

**Comment by berrnd** [comment]
It's already possible to use Authelia together with `ReverseProxyAuthMiddleware`. Authelia does the authentication and provides the authenticated username as a HTTP request header. patzly/grocy-android#255 and the there referenced issues were about that topic.

I use a lot of selfhosted things and never saw the need for centralized authentication so far. In my opinion, this potentially adds a single point of failure and a "central point of attack". In the ages of good password managers it's still convenient to have a different, lengthy, generated password I could never remember myself for every login...

**Comment by jsternadel** [comment]
Fair point.  I use a password manager just like you do.  I have no clue what 90% of my passwords even are at any given time.  It wasn't until I opened up some of my self hosted services to family that I saw the potential need for SSO.  It would be a compromise of trading central point of attack with central point of denial.  One single point to drop the ban hammer for rogue accounts.

I just pulled the repo last night and hadn't seen the `ReverseProxyAuthMiddleware` yet.  I was hoping to contribute to 2FA if I can.  I've been developing with PHP since version 4 but I've been off in C# land the last few years.  I really like your project and thought it might be something fun to get back in to PHP with.

**Comment by danionescu2007** [comment]
+1 for 2FA

**Comment by AkechiShiro** [comment]
Hi is there any draft PR or any work on this feature recently, I would like to give a it a shot as I'm interested in this feature as well for my own use.

**Comment by berrnd** [comment]
> Hi is there any draft PR or any work on this feature recently

Nothing is mentioned or referenced here, so no ([README ref](https://github.com/grocy/grocy#roadmap)).

&nbsp;
> I might need some guidance and reviews, @berrnd is that something that could be possible from the Grocy's maintainers team

The infamous "Grocy Team" was for the last 6 years, since Grocy exists, for 99.5 % (or so) me, so I can only speak for that part of the team: No, I have no time (and desire; this is a hobby project) teaching someone how to do every little detail, doing that is essentially implementing the feature in question. And of course I will review any submitted PR before merging it. If it generally works, doesn't rewrite the whole project and adapts to the current (code) style, everything has been accepted/merged so far.

**Comment by AkechiShiro** [comment]
@berrnd Good in that case, I'll figure everything on my own, if I can someday have something good, should I PR the code, if I achieve something decent or you won't have time to merge it, just asking so I know if I keep the current code style convention and all that or I go free style ?

I'm sorry if that came of like that, but I wasn't waiting for an explicit feature roadmap needed to implement 2FA on figuring out this feature, but I've also got a lot on my plate to work on, and was wondering if there would be active help or not, hence why I asked.

Also, is there any bounty system for most wanted feature request or there isn't any system like that for this hobby project ?

**Comment by berrnd** [comment]
> Also, is there any bounty system

No, there is not. This project is driven by desire (again: it's a hobby project), not money. Feel free to do/build whatever you want and please keep it from now on to the topic of the request, thanks.

**Comment by stdevel** [comment]
Are there any news on this? IMHO this would be a very useful improvement.

**Comment by berrnd** [comment]
> Are there any news on this?

Each and every tiny bit of "news" will be always visible in the corresponding issue / feature request, so here - nothing is visible here, so no progress. Also [mentioned in README](https://github.com/grocy/grocy#roadmap), spamming threads doesn't make things getting implemented faster.

**Comment by gregistech** [comment]
Wouldn't implementing SSO (Oauth2, OIDC) solve all of the possible future authentication needs?
